linked_list/implementation.py

```diff
@@ -7,34 +8,151 @@ class LinkedList(AbstractLinkedList):
     """
 
     def __init__(self, elements=None):
-        pass
+        
+        self.elements = elements
+        
+        self.start = None
+        self.end = None
+        
+        if self.elements:
+            for element in self.elements:
+                self.append(element)
 
+                
     def __str__(self):
-        pass
+        return '[{}]'.format(", ".join(str(p) for p in self.elements)) if self.elements else "[]"
+
 
     def __len__(self):
-        pass
+        return self.count()
+
 
     def __iter__(self):
-        pass
+        x = self.start
+        
+        while True:
+            yield x.elem
+            
+            if x.next == None:
+                break
+            
+            x = x.next
+            
+        raise StopIteration
+       
 
+    # THIS FUNCTION NEEDED TO RETURN A VALUE E.G. THE TEST_GET_ITEM TEST.
+    # BUT WE WERE USING IT TO RETURN A NODE IN THE POP FUNCTION.
+    # SO I WAS BREAKING THE POP WHEN I WAS MODIFYING __GETITEM__
+    # SO I CREATED THE GET_NODE FUNCTION BELOW IT. NOW EVERYTHING WORKS.
     def __getitem__(self, index):
-        pass
+        if index >= len(self):
+            raise IndexError
+
+        wantnode = self.start
+        for i in range(index): 
+            wantnode = wantnode.next
+        return wantnode.elem #ADDED .ELEM TO RETURN THE VALUE
+    
+    
+    def get_node(self, index):
+        if index >= len(self):
+            raise IndexError
+
+        wantnode = self.start
+        for i in range(index): 
+            wantnode = wantnode.next
+        return wantnode # THIS IS OUR ORIGINAL __GETITEM__ THAT RETURNS A NODE
 
+        
     def __add__(self, other):
-        pass
+        #init new list
+        n_list = LinkedList()
+        if self.elements:
+            for element in self.elements:
+                n_list.append(element)
+        
+        if other.elements:    
+            for element in other.elements:
+                n_list.append(element)
+        
```

Comment:
Here you are using elements of a container `self.elements` which you declare on init. It would have been much nicer if you use your own iterator to traverse through the nodes.